### DISCUSSION TITLE
Updating a map with dynamic properties

### DISCUSSION THREAD

**Discussion by mikejpeters** [tywalch/electrodb#455] [Q&A] [answered]
I'm trying to model a relationship between users and shops using maps with dynamic properties. I couldn't really find anything in the docs about maps with arbitrary properties, but I figured it's possible with custom types:

```ts
type ShopUsers = Record<string, { name: string; role: ShopRole }>;

const ShopEntity = new Entity({
  attributes: {
    // map of user ids to shops
    users: { type: CustomAttributeType<ShopUsers>('any') },
  },
});

type UserShops = Record<string, { name: string; role: ShopRole }>;

const UserEntity = new Entity({
  attributes: {
    // map of shop ids to users
    shops: { type: CustomAttributeType<UserShops>('any') },
  },
});
```

I thought this structure would be better than using lists, because it doesn't allow duplicate keys, and should allow me to efficiently update/remove relationships as long as I know the user id / shop id (i.e. during denormalization in a stream consumer). 

But I'm stuck actually updating/removing properties in these maps:

```ts
await ShopEntity.patch(key).set({ [`users.${userId}`]: { ... } }).go();
```
Error: Attribute "users.jsmith" does not exist on model. - For more detail on this error reference: https://electrodb.dev/en/reference/errors/#invalid-attribute

```ts
await ShopEntity.patch(key).remove([`users.${userId}`]).go();
```
TS2322: Type `users.${string}` is not assignable to type 'users'

Any help would be appreciated! Am I going in a bad direction by using custom attributes instead of lists? Or is there some other syntax to patch dynamic map properties?

**Comment by tywalch** [comment, marked as answer]
I'm AFK at the moment, but checkout the [data method](https://electrodb.dev/en/mutations/update/#data), it should allow for operations like the one you're trying to perform 👍

**Reply by mikejpeters** [reply]
Thanks so much for the quick reply! I found the data method a bit confusing at first, but really cool once I figured it out:

```ts
await ShopEntity.patch(key)
  .data(({ users }, o) => {
    o.remove(users[userId]);
  })
  .go();
  
await ShopEntity.patch(key)
  .data(({ users }, o) => {
    o.set(users[userId], { name, role });
  })
  .go();
```
